packages/flutter/lib/src/material/ink_sparkle.dart

```diff
@@ -0,0 +1,765 @@
+  ///     for clipping.
+  ///  2. If [customBorder] is null, and [borderRadius] is provided, the canvas
+  ///     is clipped by an [RRect] created from [clipCallback] and [borderRadius].
+  ///  3. If [borderRadius] is the default [BorderRadius.zero], then the [Rect] provided
+  ///      by [clipCallback] is used for clipping.
+  ///
+  /// [textDirection] is used by [customBorder] if it is non-null. This allows 
+  /// the [customBorder]'s path to be properly defined if it was the path was
+  /// expressed in terms of "start" and "end" instead of
+  /// "left" and "right".
+  ///
+  /// For examples on how the function is used, see [InkSparkle].
+  /// 
+  /// TODO(clocksmith): Refactor and extract for general ink reuse.
+  void clipCanvas({
+    required Canvas canvas,
+    required RectCallback clipCallback,
+    TextDirection? textDirection,
+    ShapeBorder? customBorder,
+    BorderRadius borderRadius = BorderRadius.zero,
+  }) {
+    final Rect rect = clipCallback();
+    if (customBorder != null) {
+      canvas.clipPath(
+          customBorder.getOuterPath(rect, textDirection: textDirection));
+    } else if (borderRadius != BorderRadius.zero) {
+      canvas.clipRRect(RRect.fromRectAndCorners(
+        rect,
+        topLeft: borderRadius.topLeft,
+        topRight: borderRadius.topRight,
+        bottomLeft: borderRadius.bottomLeft,
+        bottomRight: borderRadius.bottomRight,
+      ));
+    } else {
+      canvas.clipRect(rect);
+    }
+  }
+
+  @override
+  void paintFeature(Canvas canvas, Matrix4 transform) {
+    // Can only paint after the shader has been compiled
+    // and the _shaderManager has been initialized
+    if (_shaderManager == null) return;
+
+    final double timeMillis = _time.inMilliseconds.toDouble();
+    final double timeTouchUpMillis = _timeUpMillis.toDouble();
+
+    if (timeMillis != 0) {
+      canvas.save();
+      transformCanvas(canvas: canvas, transform: transform);
+      if (_clipCallback != null) {
+        clipCanvas(
+          canvas: canvas,
+          clipCallback: _clipCallback!,
+          textDirection: _textDirection,
+          customBorder: _customBorder,
+          borderRadius: _borderRadius,
+        );
+      }
+
+      double normalizeColorChannel(int c) => c / 255.0;
+
+      final Shader shader = _shaderManager!.shader(
+        uColor: Vector4(
+          normalizeColorChannel(_color.red),
+          normalizeColorChannel(_color.blue),
+          normalizeColorChannel(_color.green),
+          normalizeColorChannel(_color.alpha),
+        ),
+        uResolution: Vector2(referenceBox.size.width, referenceBox.size.height),
+        uTouch: Vector2(_position.dx, _position.dy),
+        uTime: timeMillis,
+        uTouchUp: timeTouchUpMillis,
+        uRippleDurationMs: 3000, // TODO(clocksmith): Confrim value
+        uPauseInflection: 0.185,
+        uPostPauseTimeScale: 1.5,
+        uPixelDensity: 1.0,
```

Comment:
these should probably be consts defined at the top with explanatory comments

Reply:
This changed since last comment. There a quite a few double values from Android 12 that do not have explanations. They were tuned by designers to make the ripple look better. Added references to those Android 12 files.